dspace-api/src/main/java/org/dspace/content/service/DSpaceObjectService.java

```diff
@@ -28,7 +28,7 @@
  * @param <T> class type
  * @author kevinvandevelde at atmire.com
  */
-public interface DSpaceObjectService<T extends DSpaceObject> extends BrowsableObjectService<T, UUID> {
+public interface DSpaceObjectService<T extends DSpaceObject> extends FindableObjectService<T, UUID> {
```

Comment:
The DSpaceObjectService should not extend the "FindableObjectService", only the ItemService, CommunityService & CollectionService should extend this one.

Reply:
It has been renamed FindableObjectService instead than Browsable or Indexable because it is not strictly related to the indexing. It is a basic interface as stated in the javadoc
https://github.com/DSpace/DSpace/pull/2391/files/5e3164bd06e2eaefc0a047087b1ab7cebdb1ee4d#diff-09f9f6d9da9fbd2c19b7ec916408eaa6R16

that should be implemented by all the persistent services that want to offer a common/standard way to retrieve a persistent object (i.e. they implement the find method).

We need to have that implemented at the DSpaceObjectService level to avoid ambiguity when we call the find method as the exact same method will be implemented in both interfaces

Reply:
As briefly discussed yesterday, the FindableObjectService has a conflict between the description, implementation and usage.

Short summary of the differences:
* The description states it's used for any persistent entity
* The implementation assumes it's only used when there's a type constant
* The actual usage assumes it's an IndexableObject

The FindableObjectService documentation states that this class can be used to find any entity by its primary key. But the class has a “getSupportsTypeConstant()” method attached to it which is not applicable for all entities.
This service is called from the following code: https://github.com/4Science/DSpace/blob/DS-4166_mydspace/dspace-api/src/main/java/org/dspace/content/factory/ContentServiceFactory.java#L111
This code returns an “IndexableObject” and indicates it only works for IndexableObject classes.

This can easily be solved by renaming the class to “IndexableObjectService”, and the T in the class should extend the “IndexableObject”. The DSpaceObjectService should no longer extend this class and only the CommunityService, CollectionService, ClaimedTaskServer, ... should extend this.

Reply:
Hi @benbosman 
> The description states it's used for any persistent entity
> The implementation assumes it's only used when there's a type constant

Is it fine to adjust the description to state that it **can be** implemented by any persistence service as I agree with you that the current description seems to suggest that it is implemented by all... to do that your persistency service of course will need to implement all the interface methods including the getSupportsTypeConstant

> The actual usage assumes it's an IndexableObject
> ..
> https://github.com/4Science/DSpace/blob/DS-4166_mydspace/dspace-api/src/main/java/org/dspace/content/factory/ContentServiceFactory.java#L111

true, the method here should be probably named getFindableObjectServiceForIndexableObject
we could have in future an use case for a generic getFindableObjectService that will be not restricted to return FindableObjectService for IndexableObject.

> This can easily be solved by renaming the class to “IndexableObjectService”, and the T in the class should extend the “IndexableObject”. The DSpaceObjectService should no longer extend this class and only the CommunityService, CollectionService, ClaimedTaskServer, ... should extend this.

Unfortunately, It is not so easy as the find method would be defined with exactly the same signature in both the IndexableObjectService than the DSpaceObjectService we will get an ambiguous reference at runtime. To solve that we will be required to add explicit cast from time to time easily to be forgotten and noted only at runtime or we should rename the find method in the IndexableObjectService in findIndexableObject to avoid conflict and map from one to the other in the Item/Community/Collection... services

Are you fine with
1) rephrasing the description and rename the getFindableObjectService
**or**
2) do you want to go with an IndexableObjectService that will use a different name for the find method?

Reply:
@abollini and @benbosman :  I don't have a strong opinion either way on @abollini's two options, but it does seem like this `FindableObjectService` is not able to be implemented quite as described (at least not at this time).  So, it makes me wonder if we should go with @abollini's option 2, and rename it to be `IndexableObjectService` and change the `find()` method to be named `findIndexableObject()`.  I know that @abollini was trying to build in a more generic (broadly adoptable) "find" service, but maybe that should come in a future refactor (as it has little to do with the MyDSpace feature).

Reply:
Hi @abollini I understand the problem with the ambiguous reference, which will indeed require a larger refactor.
As suggested by @tdonohue above, given the time constraints, I agree we can go with @abollini's option 2, and rename it to be `IndexableObjectService` and change the `find()` method to be named `findIndexableObject()`.

Reply:
I have implemented and pushed the option 2. Please note that there is also a new getSupportsIndexableObjectType method for the same reason explained above for the find method. I put a note about that also in the javadoc of the IndexableObjectService and its implementation so to remember us in future why this "strange" methods alias are here